### ISSUE TITLE
SM2/SM3/SM4 algorithm support for ssl protocol

### ISSUE THREAD

**Issue by Yikun** [openssl/openssl#12473]
Hi, team

We noticed that SM2/SM3/SM4 has been added in openssl.[1][2][3]

But the the SM2/3/4 has not been supported in ssl protocol[4], that means we could only use local API to use the SM algorithm.

Do we have any plan to support SM algorithm in ssl protocol?

[1] https://github.com/openssl/openssl/pull/4793
[2] https://github.com/openssl/openssl/pull/4616
[3] https://github.com/openssl/openssl/pull/4552

**Comment by InfoHunter** [comment]
If the word 'SSL protocol' here stands for the Chinese "GM/T 0024" standards (which defines a dual certificate modified TLS 1.1 protocol), then we don't have a plan to support it yet.

**Comment by Yikun** [comment]
@InfoHunter Thanks for you reply.

Yes, we are talking about the same thing.

So, Does the community have some discussion about this? Is it possible to support this feature in recent releases?

**Comment by InfoHunter** [comment]
Actually I recall that we have discussed this a bit on last October's committers' meeting in Germany. But considering the complexity GM/T 0024 brings to the TLSv1.1 protocol and the timeline of OpenSSL 3.0, it's not a high priority task. So I am not optimistic for including such a big feature in recent releases of OpenSSL - but we could discuss again on this topic after OpenSSL 3.0 is released.

BTW, what I personally prefer to bring into OpenSSL is using SM algorithms in standard TLS protocol, say TLS 1.3 for instance. I made a draft last year at: https://datatracker.ietf.org/doc/html/draft-yang-tls-tls13-sm-suites-04. Don't know if this fulfills your requirement?

**Comment by paulidale** [comment]
For OpenSSL 3.0, I don't think that there is any chance GM/T 0024 would be included.  After, 3.0 it might be a possibility.  It meets the minimum standard for inclusion in OpenSSL (national standard) but would require a decision to include it and somebody to do the work.  Would it still make sense in light of @InfoHunter's draft for TLS 1.3?

I'm kind of expecting some protocol discussions to happen after 3.0 stabilises.

**Comment by InfoHunter** [comment]
> I'm kind of expecting some protocol discussions to happen after 3.0 stabilises.

Agree. But one forecast: the GM/T 0024 is very ambiguous to implement, we really need to take a lot of care when dealing with this protocol ;-)

**Comment by romen** [comment]
I don't know the details about the protocol use of this, but as a related problem it would be beneficial if someone could send string feedback to the Chinese standardization entity in charge of these algorithms highlighting how integration would actually be easier if OIDs were issued and mandated for serializing objects.

@InfoHunter is there any option of being proactive on that?

**Comment by InfoHunter** [comment]
Ha, then I think you found the right person to talk to ;-). I could probably do what you proposed at a right time.

**Comment by InfoHunter** [comment]
But I definitely need a remainder of this - I think I will dive into this months later when the time comes...

**Comment by romen** [comment]
> But I definitely need a remainder of this - I think I will dive into this months later when the time comes...

😜 Can you remind me to remind you?

**Comment by WordlessEcho** [comment]
May I ask any new details about this issue?

**Comment by FlyGoat** [comment]
Hi, @InfoHunter I saw Alipay had implement RFC8998 in your OpenSSL fork https://github.com/Tongsuo-Project/Tongsuo, are you going to upstream those changes, or at least authorise third-party to upstream those changes based on your work?

Thanks.

**Comment by youyang-hub** [comment]
Is there any interest on taking TLS_SM4_GCM_SM3 implementations into OpenSSL repo?

**Comment by mattcaswell** [comment]
> Is there any interest on taking TLS_SM4_GCM_SM3 implementations into OpenSSL repo?

It's not currently on our list of things for implementation, but I think we would definitely consider an external PR contributing support for this.

I'd be interested in @InfoHunter's input on this since he wrote the RFC for that ciphersuite!

**Comment by shove70** [comment]
What's new about this issue？ ：）

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 3 months otherwise.

**Comment by mbroz** [comment]
For TLS1.3 SM cipher suites, see #29953 and #30028.

**Comment by t8m** [comment]
This is implemented in 4.0

**Comment by shove70** [comment]
> This is implemented in 4.0

That's really good news
